libvirt.go

```diff
@@ -553,23 +540,45 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
 }
 
 func (d *Driver) GetIP() (string, error) {
-	log.Debugf("GetIP called for %s", d.MachineName)
-	mac, err := d.getMAC()
+	s, err := d.GetState()
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("%v : machine in unknown state", err)
+	}
+	if s != state.Running {
+		return "", errors.New("host is not running")
+	}
+	ip, err := d.lookupIP()
+	if err != nil {
+		return "", fmt.Errorf("%v : getting IP", err)
+	}
+
+	return ip, nil
+}
+
+func (d *Driver) lookupIP() (string, error) {
+	conn, err := d.getConn()
+	if err != nil {
+		return "", fmt.Errorf("%v : getting connection and domain", err)
+	}
+
+	defer conn.Close()
+
+	libVersion, err := conn.GetLibVersion()
+	if err != nil {
+		return "", fmt.Errorf("%v: getting libversion", err)
 	}
-	/*
-	 * TODO - Figure out what version of libvirt changed behavior and
-	 *        be smarter about selecting which algorithm to use
-	 */
-	ip, err := d.getIPByMACFromLeaseFile(mac)
-	if ip == "" {
-		ip, err = d.getIPByMacFromSettings(mac)
+	log.Debugf("Libvirt version: %s", libVersion)
+
+	mac, err := d.getMAC()
+	if err != nil {
+		return "", fmt.Errorf("%v : Not able to get mac address", err)
 	}
 
-	return ip, err
+	// TODO: for everything > 1002006, there is direct support in the libvirt-go for handling this
+	return d.getIPByMacFromSettings(mac)
```

Comment:
libvirt-go is still parsing some internal libvirt files to achieve this. I believe the libvirt C library offers an API to directly get the IP rather than having to do it ourselves in libvirt.go

Reply:
@cfergeau I don't have time right now to check in the C library implementation part and check if libvirt-go wrapper have same functionality. I can put a TODO here and we can circle back to it if that makes sense?